### PR TITLE
refactor to make callable interface

### PR DESCRIPTION
**Pull request by PeterDSteinberg**
Installing this PR:
 * Run `python build.py`
 * Find the target/release build output folder
 * Make an environment variable in your ~/.bashrc: `export WHITEBOX_TOOLS_BUILD=/Users/psteinberg/Documents/earth/whitebox-geospatial-analysis-tools/whitebox_tools/target/release`
 * Run `python setup.py develop`

What does this PR do:
 * [x] Makes `whitebox_tools` installable with a `setup.py`
 * [x] Makes a Python callable interface to the WhiteBox CLI tools, e.g.
```
import os
from whitebox_tools import Aspect

a = Aspect(input=os.path.expanduser('~/Documents/earth/whitebox-geospatial-analysis-tools/whitebox_tools/whitebox_tools/tests/testdata/DEM.dep'), output=os.path.abspath('aspect2.dep'), zfactor=1)
```
 * [x] Makes CLI tools for each WhiteBox tool so that `-r=` arg does not have to be passed in, e.g.:
```
whitebox-Aspect -i `pwd`/whitebox_tools/tests/testdata/DEM.dep -o `pwd`/aspect.dep --zfactor 1.0
```

This is a work in progress.

TODO:
 * [ ] Use `param`
 * [x] Currently paths passed into new CLI tools and functions (examples above) require absolute paths.  It should be a quick change to convert all paths to absolute ones.

### CODE REVIEW

File: whitebox_tools/build.py [hunk 1]
```diff
@@ -1,9 +1,14 @@
...
 ''' this module is used to build whitebox-tools.
 '''
+from __future__ import print_function
 import os
 import sys
-from subprocess import call
+from subprocess import call as sp_call
+
+def call(*args, **kw):
+    print('Running: ', args[0], '(with kwargs = {})'.format(kw))
```
Comment (PeterDSteinberg) [inline whitebox_tools/build.py]: print the command before running it so user can see if it fails, e.g. `cargo` / rust were not installed

File: whitebox_tools/build.py [hunk 2]
```diff
@@ -23,18 +28,16 @@ def main():
             # Update #
             retcode = call(['cargo', 'update'], shell=False)
             if retcode < 0:
-                print >>sys.stderr, "Child was terminated by signal", -retcode
+                print( "Child was terminated by signal", -retcode, file=sys.stderr)
```
Comment (PeterDSteinberg) [inline whitebox_tools/build.py]: Python 3 compatibility

File: whitebox_tools/examples/whitebox_example.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python
+''' This module provides examples of how to call the whitebox_tool script and the
+whitebox-tools geospatial analysis library using Python code.
+'''
+from __future__ import print_function
+from functools import partial
+import os
+import sys
+from whitebox_cli import make_wbt
```
Comment (PeterDSteinberg) [inline whitebox_tools/examples/whitebox_example.py]: TODO: I need to update this script.  This import statement likely doesn't work.  It should be just `from whitebox_tools import WhiteBox` then later `wbt = WhiteBox()`

File: whitebox_tools/setup.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import versioneer
+
+version = versioneer.get_version()
+cmdclass = versioneer.get_cmdclass()
+
+def make_console_scripts():
+    import whitebox_tools as wt
+    scripts = []
+    for tool in wt.tools:
+        line = 'whitebox-{0} = whitebox_tools:{0}Cli'.format(tool)
```
Comment (PeterDSteinberg) [inline whitebox_tools/setup.py]: Get the `*Cli` callables, e.g.`whitebox-Aspect = whitebox_tools:AspectCli` - which  as a console entry point analogous to `Aspect` (See usage example of `Aspect` in PR's first comment)

File: whitebox_tools/setup.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+setup(name='whitebox_tools',
+      version=version,
+      cmdclass=cmdclass,
+      description='WhiteBox Tools',
+      include_package_data=True,
+      install_requires=[],
+      packages=find_packages(),
+      package_data={},
+      entry_points={
+        'console_scripts': make_console_scripts(), #['whitebox_tools = whitebox_tools.whitebox_cli:call_whitebox_cli'],
```
Comment (PeterDSteinberg) [inline whitebox_tools/setup.py]: TODO: I need to remove the commented part

File: whitebox_tools/whitebox_example.py [hunk 6]
```diff
@@ -1,98 +0,0 @@
...
-        # Run the tool and check the return value
-        if wbt.run_tool(name, args, callback) != 0:
-            print("ERROR running {}".format(name))
-
-    except:
-        print("Unexpected error:", sys.exc_info()[0])
-        raise
-
-
-def callback(out_str):
```
Comment (PeterDSteinberg) [inline whitebox_tools/whitebox_example.py]: `callback` was moved to `whitebox_tools.whitebox_cli`

File: whitebox_tools/whitebox_tools/whitebox_base.py [hunk 7]
```diff
@@ -23,23 +23,29 @@ class WhiteboxTools(object):
...
+            exe_path = os.path.join(target, 'release')
+            if not exe_path or not os.path.exists(exe_path):
+                raise ValueError('Define WHITEBOX_TOOLS_BUILD environment variable')
+
+        self.exe_path = exe_path
+        if platform == 'win32':
+            exe = "whitebox_tools.exe"
+        else:
+            exe = 'whitebox_tools'
+        self.exe_name = os.path.join(self.exe_path, exe)
```
Comment (PeterDSteinberg) [inline whitebox_tools/whitebox_tools/whitebox_base.py]: Detecting the `target/release` build dir.  Use environment variable: `WHITEBOX_TOOLS_BUILD`

File: whitebox_tools/whitebox_tools/whitebox_base.py [hunk 8]
```diff
@@ -51,16 +57,40 @@ def set_verbose_mode(self, val=True):
...
+            sys.stdout.flush()
+            if line != '':
+                if not self.cancel_op and not silent:
+                    callback(line.strip())
+                else:
+                    self.cancel_op = False
+                    proc.terminate()
+                    ret_code = 2
+            else:
+                break
```
Comment (PeterDSteinberg) [inline whitebox_tools/whitebox_tools/whitebox_base.py]: This section of subprocess management was repeated in `WhiteboxTools`

File: whitebox_tools/whitebox_tools/whitebox_cli.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        parsers[tool] = convert_help_extract_params(tool, wbt)
+    return parsers
+
+
+for tool in tools:
+    globals()[tool + 'Cli'] = partial(call_whitebox_cli, tool)
+    globals()[tool] = partial(call_whitebox_func, tool)
+
+extras = [t + 'Cli' for t in tools]
+__all__ = tools + ['callback', 'tools', 'WhiteboxTools'] + extras
```
Comment (PeterDSteinberg) [inline whitebox_tools/whitebox_tools/whitebox_cli.py]: add `get_all_parsers` to `__all__` or remove that function

**Comment by dharhas** [comment]
+1 on accepting relative paths and converting to absolute before passing to whitebox tools

**Comment by jblindsay** [comment]
Hi Peter,
You've made some rather large and fundamental changes to the code and I'm afraid that I'm having a difficult time interpreting the changes without context. Can you please email me to give me some idea of what everything that you've done actually does. I'm reluctant to merge these changes without an understanding of what is going on.

**Comment by PeterDSteinberg** [comment]
Hi @jblindsay - The basic idea of the PR is to wrap the `WhiteBox` tools for usage in interactive Python, passing `xarray.Dataset` or `xarray.DataArray` objects as the `--input` argument to the `Whitebox` tools.  The aim is to allow Python scripts to pass `xarray` data structures in and out of `WhiteBox` tools like `Aspect`.  I'll follow up privately with you.  I'm happy to discuss further.  One option is to not merge this PR for a while (closing it now), and to leave development in the `ContinuumIO` fork for 1-2 months while it is being refined.  Then revisit the discussion about merging.  I agree it is a lot of changes and in need of simplification.

**Comment by jblindsay** [comment]
Hi Peter,

By adding xarray, have you not just created a numpy dependency to the project? That is a fairly major dependency. Perhaps I'm missing something.

**Pull request closed without merging**